pennylane/templates/layers/custom_entangler.py

```diff
@@ -0,0 +1,262 @@
+
+        The :mod:`~pennylane.init` module has two parameter initialization functions, ``basic_entangler_layers_normal``
+        and ``basic_entangler_layers_uniform``.
+
+        .. code-block:: python
+
+            from pennylane.init import basic_entangler_layers_normal
+
+            n_layers = 4
+            weights = basic_entangler_layers_normal(n_layers=n_layers, n_wires=n_wires)
+
+            circuit(weights)
+
+
+        **No periodic boundary for two wires**
+
+        When using two wires, the convention is to drop the periodic boundary condition.
+        This means that the connection from the second to the first wire is omitted.
+
+        .. code-block:: python
+
+            n_wires = 2
+            dev = qml.device('default.qubit', wires=n_wires)
+
+            @qml.qnode(dev)
+            def circuit(weights):
+                BasicEntanglerLayers(weights=weights, wires=range(n_wires))
+                return [qml.expval(qml.PauliZ(wires=i)) for i in range(n_wires)]
+
+        >>> circuit([[pi, pi]])
+        [-1, 1]
+
+
+        **Changing the rotation gate**
+
+        Any single-qubit gate can be used as a rotation gate, as long as it only takes a single parameter. The default is the ``RX`` gate.
+
+        .. code-block:: python
+
+            @qml.qnode(dev)
+            def circuit(weights):
+                BasicEntanglerLayers(weights=weights, wires=range(n_wires), rotation=qml.RZ)
+                return [qml.expval(qml.PauliZ(wires=i)) for i in range(n_wires)]
+
+        Accidentally using a gate that expects more parameters throws a
+        ``ValueError: Wrong number of parameters``.
+
+        **Using the `interactions` argument**
+
+        By using `interactions`, and custom broadcasting pattern of CNOT gates can be placed on the circuit.
+
+        .. code-block:: python
+
+            n_wires = 4
+            dev = qml.device('default.qubit', wires=n_wires)
+            interactions = [[0, 1], [2, 3]]
+
+            @qml.qnode(dev)
+            def circuit(weights):
+                BasicEntanglerLayers(weights=weights, wires=range(n_wires), rotation=qml.RZ, interactions)
+                return [qml.expval(qml.PauliZ(wires=i)) for i in range(n_wires)]
+
+        >>> circuit([[pi, pi, pi, pi]])
+        [-1, 1, -1, 1]
+    """
+
+    #############
+    # Input checks
+
+    n_parameters = {
+        "single": len(wires),
+        "double": 0 if len(wires) in [0, 1] else len(wires) // 2,
+        "double_odd": 0 if len(wires) in [0, 1] else (len(wires) - 1) // 2,
+        "chain": 0 if len(wires) in [0, 1] else len(wires) - 1,
+        "ring": 0 if len(wires) in [0, 1] else (1 if len(wires) == 2 else len(wires)),
+        "pyramid": 0 if len(wires) in [0, 1] else sum(i + 1 for i in range(len(wires) // 2)),
+        "all_to_all": 0 if len(wires) in [0, 1] else len(wires) * (len(wires) - 1) // 2,
+    }
+
+    OPTIONS = ["single", "double", "double_odd", "chain", "ring", "pyramid", "all_to_all"]
```

Comment:
Does it make more sense to import options from the broadcast function? That way it's more maintainable

Reply:
Yeah, this definitely makes more sense, although the OPTIONS and n_parameters variables are contained in the broadcast method, so is it possible to access them from importing braodcast.py?

Reply:
yep, that's what I had in mind

if the are inside the method, it's relatively simple to move them out to the larger namespace

Reply:
Ok, done!

Reply:
I added a new method to braodcast.py that returns the object of param numbers for each pattern (for an inputted number of wires), which I use in custom_entangler.py

Reply:
Just as a quick glance, that looks suitable, but I'll defer to @josh146 and @mariaschuld for the final code review

Reply:
@co9olguy Sounds good, thanks!

Reply:
lol do you have these notifications hard-wired into your cerebral cortex? you're always so fast

Reply:
Surprisingly no. Although it's pre-historic compared to Internet implanted in the brain, email notifications work pretty well 😅

Reply:
I had to turn those off a long-time ago :grinning:, way too many to sort through